### PR TITLE
[Dynamics] Drone dynamics

### PR DESCRIPTION
**Pull request by gy2256**
The dynamics require testing. Flying this manually is challenging because the action contains [Total_thurst, Torque_x, Torque_y, Torque_z]. I will code a separate LQR controller (as a heuristic controller) for testing.

### CODE REVIEW

File: vmas/simulator/dynamics/drone.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        self.I_zz = I_zz
+        self.world = world
+        self.g = 9.81
+        # Drone state: phi(roll), theta (pitch), psi (yaw),
+        #              p (roll_rate), q (pitch_rate), r (yaw_rate),
+        #              x_dot (vel_x), y_dot (vel_y), z_dot (vel_z),
+        #              x (pos_x), y (pos_y), z (pos_z)
+        self.drone_state = torch.tensor(
+            [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0]
+        )
```
Comment (matteobettini) [inline vmas/simulator/dynamics/drone.py:38]: So this should have the device of the world and the first batched dimension.

You can do it with 
```python
self.drone_state = torch.tensor(
            [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0], device=world.device
        ).unsqueeze(0).expand(world.batch_dim, 12)
```

File: vmas/simulator/dynamics/drone.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
+from typing import Union
+
+import torch
+import vmas.simulator.core
+import vmas.simulator.utils
+from vmas.simulator.dynamics.common import Dynamics
+
+
+class Drone(Dynamics):
```
Comment (matteobettini) [inline vmas/simulator/dynamics/drone.py:9]: You need to implement the reset function

File: vmas/simulator/dynamics/drone.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        k2 = f(state + self.dt * k1 / 2)
+        k3 = f(state + self.dt * k2 / 2)
+        k4 = f(state + self.dt * k3)
+        return state + (self.dt / 6) * (k1 + 2 * k2 + 2 * k3 + k4)
+
+    def needed_action_size(self) -> int:
+        return 4
+
+    def process_action(self):
+        # Extracts the velocity and steering angle from the agent's actions and convert them to physical force and torque
```
Comment (matteobettini) [inline vmas/simulator/dynamics/drone.py:54]:
```suggestion
      
```

File: vmas/simulator/dynamics/drone.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        u = self.agent.action.u
+        thrust = u[:, 0]  # Thrust, sum of all propeller thrusts
+        torque = u[:, 1:4]  # Torque in x, y, z direction
+
+        thrust += self.mass * self.g  # Ensure the drone is not falling
```
Comment (matteobettini) [inline vmas/simulator/dynamics/drone.py:59]: mass should come from `self.agent.mass` you can remove it from the constructor

File: vmas/simulator/dynamics/drone.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            s_theta = torch.sin(theta)
+            c_psi = torch.cos(psi)
+            s_psi = torch.sin(psi)
+
+            # Postion Dynamics
+            x_ddot = (c_phi * s_theta * c_psi + s_phi * s_psi) * thrust / self.mass
+            y_ddot = (c_phi * s_theta * s_psi - s_phi * c_psi) * thrust / self.mass
+            z_ddot = (c_phi * c_theta) * thrust / self.mass - self.g
+            # Angular velocity dynamics
+            p_dot = (torque[0] - (self.Iyy - self.Izz) * q * r) / self.Ixx
```
Comment (matteobettini) [inline vmas/simulator/dynamics/drone.py:83]: `torque[0]`: the first dim is the number of enviornments, so you might want `torque[:,0]`. Applies to other points as well

File: vmas/simulator/dynamics/drone.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            s_phi = torch.sin(phi)
+            c_theta = torch.cos(theta)
+            s_theta = torch.sin(theta)
+            c_psi = torch.cos(psi)
+            s_psi = torch.sin(psi)
+
+            # Postion Dynamics
+            x_ddot = (c_phi * s_theta * c_psi + s_phi * s_psi) * thrust / self.mass
+            y_ddot = (c_phi * s_theta * s_psi - s_phi * c_psi) * thrust / self.mass
+            z_ddot = (c_phi * c_theta) * thrust / self.mass - self.g
```
Comment (matteobettini) [inline vmas/simulator/dynamics/drone.py:81]: same mass thing

**Review by matteobettini — COMMENTED** [review]
Very very cool, here a first round of comments

File: vmas/simulator/dynamics/drone.py (hunk 6 above)
Comment (gy2256) [reply vmas/simulator/dynamics/drone.py]: Fixed

File: vmas/simulator/dynamics/drone.py (hunk 2 above)
Comment (matteobettini) [reply vmas/simulator/dynamics/drone.py:9]: In that function you should reset the internal state of the controller to 0s , but do not touch the vmas state.
This is why I would keep the 2 states separate and cat them only in the process_action function.

The function will take an `env_index` (look at parent class).
What you need to do is:
- when the index is None, reset all envs
- when the index is int, reset only the environment at that index

Comment (gy2256) [reply vmas/simulator/dynamics/drone.py:9]: Also, I'm thinking of the correct way to trigger the reset function. For example, the drone has flipped over when some internal state (pitch, roll) goes over some threshold. Thus, the reset() function should be called. Can this be included during training?

Comment (matteobettini) [reply vmas/simulator/dynamics/drone.py:9]: the reset function is automatically handled by vmas. People can decide when to trigger it based on the scenario. For now let's not think about this

Comment (matteobettini) [reply vmas/simulator/dynamics/drone.py:9]: see here for example https://github.com/proroklab/VectorizedMultiAgentSimulator/blob/a883c90248f264136cc1840c911da96c87d5a4d7/vmas/simulator/controllers/velocity_controller.py#L80

Comment (gy2256) [reply vmas/simulator/dynamics/drone.py:9]: Thanks. I think I got it. I will double check with you tomorrow in-person before updating the code.

File: vmas/simulator/dynamics/drone.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    def process_action(self):
+        u = self.agent.action.u
+        thrust = u[:, 0]  # Thrust, sum of all propeller thrusts
+        torque = u[:, 1:4]  # Torque in x, y, z direction
+
+        thrust += self.agent.mass * self.g  # Ensure the drone is not falling
+
+        vmas_state = torch.cat(
+            (self.agent.state.pos, self.agent.state.rot), dim=1
+        )  # (x,y,rotation)
```
Comment (matteobettini) [inline vmas/simulator/dynamics/drone.py:67]: can we avoid this cat since then we use the individual elements?

**Review by matteobettini — APPROVED** [review]
LGTM.

Did you format it already?

As we discussed it would be useful to have a function `needs_resetting` that returns a boolean tensor of shape `n_envs` telling if the drone is in an unrecoverable state for each env

We should have a debug scenario for this (like for diff drive) where users can test it.
Maybe this will need a simple controller layer.

File: vmas/simulator/dynamics/drone.py (hunk 7 above)
Comment (gy2256) [reply vmas/simulator/dynamics/drone.py]: yeah. The vmas_state is removed.

**Comment by gy2256** [comment]
I will add needs_resetting. 

Speaking of the debugging scenario, I also want to manually fly the drone with the raw control. I can also provide an LQR controller for debugging, where a user inputs translational velocities. 

Is the following code a good place to start?
[https://github.com/proroklab/VectorizedMultiAgentSimulator/blob/main/vmas/interactive_rendering.py](https://github.com/proroklab/VectorizedMultiAgentSimulator/blob/main/vmas/interactive_rendering.py)

**Comment by matteobettini** [comment]
> I will add needs_resetting.
> 
> Speaking of the debugging scenario, I also want to manually fly the drone with the raw control inputs [thrust, torque_x, torque_y, torque_z]. Then, I can provide an LQR controller for debugging, where a user inputs translational velocities.
> 
> Is the following code a good place to start? https://github.com/proroklab/VectorizedMultiAgentSimulator/blob/main/vmas/interactive_rendering.py

Yes, executting a scenario with a drone and that will let you control the first 3 actions (action one with left/right, action 2 with up/down, action 3 with m/n). (This buttons will do -/+ max_range whioch you set upon agent construction)

In your case you have 4 actions so we might need a workaround for that.

Regarding the controller, look at `velocity_controller`. You can make a similar one that it is called in `process_action` before the dynamics. 

We can decide if we want the input of this to be forces and torque or velocities.

File: vmas/simulator/dynamics/drone.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+            ).unsqueeze(0).repeat(self.world.batch_dim, 1)
+        else:
+            self.drone_state[index] = 0.0
+            
+    def need_reset(self) -> bool:
+        # Constraint roll and pitch within +-30 degrees
+        if torch.any(torch.abs(self.drone_state[:, 0:2]) > 30 * (torch.pi / 180)):
+            return True
+
+        return False
```
Comment (matteobettini) [inline vmas/simulator/dynamics/drone.py:52]:
```suggestion
    def need_reset(self) -> bool:
        # Constraint roll and pitch within +-30 degrees
        return torch.any(self.drone_state[:, :2].abs() > 30 * (torch.pi / 180), dim=-1)
 
```

You want to reset if the abs value of any of the first 2 dims is grater than that right?

File: vmas/simulator/dynamics/drone.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        self.reset()
+        
+    def reset(self, index: Union[Tensor, int] = None):
+        if index is None:
+            # Drone state: phi(roll), theta (pitch), psi (yaw),
+            #              p (roll_rate), q (pitch_rate), r (yaw_rate),
+            #              x_dot (vel_x), y_dot (vel_y), z_dot (vel_z),
+            #              x (pos_x), y (pos_y), z (pos_z)
+            self.drone_state = torch.tensor(
+                [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], device=self.world.device
```
Comment (matteobettini) [inline vmas/simulator/dynamics/drone.py:42]: these are all zeros now?

If so you can do `torch.zeros(self.world.batch_dim,12,device-self.world.device)`

File: vmas/simulator/dynamics/drone.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        def f(state):
+            phi = state[:, 0]
+            theta = state[:, 1]
+            psi = state[:, 2]
+            p = state[:, 3]
+            q = state[:, 4]
+            r = state[:, 5]
+            x_dot = state[:, 6]
+            y_dot = state[:, 7]
+            z_dot = state[:, 8]
```
Comment (matteobettini) [inline vmas/simulator/dynamics/drone.py:87]: dumb question but the states 9,10,11 are not used?

File: vmas/simulator/dynamics/drone.py [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+            
+
+        if self.integration == "euler":
+            new_drone_state = self.euler(f, self.drone_state)
+        else:
+            new_drone_state = self.runge_kutta(f, self.drone_state)
+
+        # Calculate the change in state
+        delta_state = new_drone_state - self.drone_state
+        self.drone_state[: ,] = new_drone_state
```
Comment (matteobettini) [inline vmas/simulator/dynamics/drone.py:131]:
```suggestion
        self.drone_state = new_drone_state
```

File: vmas/simulator/dynamics/drone.py (hunk 10 above)
Comment (gy2256) [reply vmas/simulator/dynamics/drone.py:92]: You're correct because the time derivative of the drone state does not depend on the translational position, so we do not use states 9, 10, and 11.

File: vmas/simulator/dynamics/drone.py (hunk 9 above)
Comment (gy2256) [reply vmas/simulator/dynamics/drone.py]: Yeah, because the initial height (the last element in the state) can be zero due to the lack of "ground" in VMAS. Also, it's easier to do the reset function. We allow the drone to fly at a negative altitude. We might need to think of a way to define the spatial constraint in the z-direction. (for need_reset() function or training in general).

File: vmas/simulator/dynamics/drone.py (hunk 8 above)
Comment (gy2256) [reply vmas/simulator/dynamics/drone.py]: Yeah. The idea to constraint the roll or pitch angle of the drone between -30 and 30 degrees, so the drone will not flip over.
